File: app/src/main/java/top/suto/appopt/RuleSyntax.kt
package top.suto.appopt

/** 将 applist.conf 的全部受支持语法展开为统一规则，其他功能只处理这一种内部语义。 */
object RuleSyntax {

    enum class Format {
        STANDARD,
        TAGGED,
        NATURAL,
        NESTED,
        FUNCTION,
        YAML
    }

    data class Rule(
        val owner: String,
        val thread: String?,
        val cpus: String
    ) {
        val canonicalLine: String
            get() = "${thread?.let { "$owner{$it}" } ?: owner}=$cpus"
    }

    data class Segment(
        val rawLines: List<String>,
        val rules: List<Rule>,
        val ownerHint: String? = null,
        val block: Boolean = false,
        val valid: Boolean = true,
        val format: Format? = null
    )

    data class Document(val segments: List<Segment>) {
        val rules: List<Rule> = segments.flatMap(Segment::rules)
    }

    private data class Header(
        val owner: String,
        val fallback: String?,
        val format: Format,
        val valid: Boolean = true
    )

    fun parse(text: String): Document {
        val lines = text.lines()
        val segments = mutableListOf<Segment>()
        var index = 0
        while (index < lines.size) {
            val raw = lines[index]
            val code = codePart(raw)
            if (code.isEmpty() || code.startsWith("#")) {
                segments += Segment(listOf(raw), emptyList())
                index++
                continue
            }

            val yamlHeader = parseYamlHeader(raw)
            if (yamlHeader != null) {
                val end = yamlEnd(lines, index)
                val parsed = parseYamlBody(lines.subList(index + 1, end), yamlHeader.owner)
                val segmentValid = yamlHeader.valid && parsed != null
                segments += Segment(
                    rawLines = lines.subList(index, end),
                    rules = if (segmentValid) parsed.orEmpty() else emptyList(),
                    ownerHint = yamlHeader.owner,
                    block = true,
                    valid = segmentValid,
                    format = Format.YAML
                )
                index = end
                continue
            }

            val header = parseBraceHeader(code)
            if (header != null) {
                val end = braceBlockEnd(lines, index)
                if (end == null) {
                    segments += Segment(
                        rawLines = lines.subList(index, lines.size),
                        rules = emptyList(),
                        ownerHint = header.owner,
                        block = true,
                        valid = false,
                        format = header.format
                    )
                    break
                }
                val close = codePart(lines[end - 1])
                val resolvedHeader = resolveTailFallback(header, close)
                val body = lines.subList(index + 1, end - 1)
                val nested = resolvedHeader.format == Format.TAGGED && body.any {
                    codePart(it) == "threads {" || codePart(it) == "processes {"
                }
                val effectiveHeader = if (nested) resolvedHeader.copy(format = Format.NESTED) else resolvedHeader
                val parsed = when (effectiveHeader.format) {
                    Format.NESTED -> parseNestedBody(body, effectiveHeader)
                    else -> parseBraceBody(body, effectiveHeader)
                }
                val segmentValid = effectiveHeader.valid && parsed != null
                segments += Segment(
                    rawLines = lines.subList(index, end),
                    rules = if (segmentValid) parsed.orEmpty() else emptyList(),
                    ownerHint = effectiveHeader.owner,
                    block = true,
                    valid = segmentValid,
                    format = effectiveHeader.format
                )
                index = end
                continue
            }

            val legacy = parseLegacyRule(code)
            segments += Segment(
                rawLines = listOf(raw),
                rules = legacy?.let(::listOf).orEmpty(),
                ownerHint = legacy?.owner,
                valid = legacy != null
            )
            index++
        }
        return Document(segments)
    }

    fun parseLegacyRule(rawLine: String): Rule? {
        val code = codePart(rawLine)
        val eq = code.indexOf('=')
        if (eq <= 0) return null
        val key = code.substring(0, eq).trim()
        val cpus = code.substring(eq + 1).trim()
        if (key.isEmpty() || cpus.isEmpty()) return null
        val open = key.indexOf('{')
        if (open < 0) {
            if ('}' in key) return null
            return Rule(key, null, cpus)
        }
        val close = key.lastIndexOf('}')
        if (open == 0 || close != key.lastIndex || close <= open + 1 || key.indexOf('{', open + 1) >= 0) {
            return null
        }
        val owner = key.substring(0, open).trim()
        val thread = key.substring(open + 1, close).trim()
        return if (owner.isEmpty() || thread.isEmpty()) null else Rule(owner, thread, cpus)
    }

    private fun parseBraceHeader(code: String): Header? {
        if (!code.endsWith('{')) return null
        val prefix = code.dropLast(1).trim()
        if (prefix == "app" || prefix.startsWith("app ")) {
            val match = NATURAL_HEADER.matchEntire(code)
                ?: return Header("", null, Format.NATURAL, valid = false)
            return Header(match.groupValues[1], match.groupValues[2].ifEmpty { null }, Format.NATURAL)
        }
        if (prefix.startsWith("app(")) {
            val args = prefix.removePrefix("app(").removeSuffix(")")
            if (!prefix.endsWith(')')) return Header("", null, Format.FUNCTION, valid = false)
            val values = args.split(',').map(String::trim)
            val owner = values.firstOrNull().orEmpty()
            val fallback = values.getOrNull(1)
            return Header(
                owner,
                fallback,
                Format.FUNCTION,
                owner.isNotEmpty() && values.size in 1..2 && values.none(String::isEmpty)
            )
        }
        if (prefix.isEmpty() || prefix.contains('}')) return null
        val eq = prefix.indexOf('=')
        if (eq >= 0) {
            val owner = prefix.substring(0, eq).trim()
            val fallback = prefix.substring(eq + 1).trim()
            return if (fallback.isEmpty()) {
                Header(owner, null, Format.TAGGED, owner.isNotEmpty())
            } else {
                Header(owner, fallback, Format.STANDARD, owner.isNotEmpty())
            }
        }
        return Header(prefix, null, Format.STANDARD, prefix.isNotEmpty())
    }

    private fun resolveTailFallback(header: Header, close: String): Header {
        if (header.format != Format.STANDARD) {
            return header.copy(valid = header.valid && close == "}")
        }
        val tail = close.removePrefix("}").trim()
        if (tail.isEmpty()) return header
        val fallback = tail.removePrefix("=").trim()
        return header.copy(
            fallback = fallback.takeIf { it.isNotEmpty() && header.fallback == null },
            valid = header.valid && tail.startsWith('=') && fallback.isNotEmpty() && header.fallback == null
        )
    }

    private fun parseBraceBody(lines: List<String>, header: Header): List<Rule>? {
        val rules = mutableListOf<Rule>()
        var bodyFallback = false
        for (raw in lines) {
            val code = codePart(raw)
            if (code.isEmpty() || code.startsWith("#")) continue
            val rule = when (header.format) {
                Format.STANDARD -> parseStandardMember(header.owner, code)
                Format.TAGGED -> parseTaggedMember(header.owner, code)
                Format.NATURAL -> parseNaturalMember(header.owner, code)
                Format.FUNCTION -> parseFunctionMember(header.owner, code)
                Format.NESTED, Format.YAML -> null
            } ?: return null
            if (header.format == Format.TAGGED && rule.owner == header.owner && rule.thread == null) {
                if (bodyFallback || header.fallback != null) return null
                bodyFallback = true
            }
            rules += rule
        }
        header.fallback?.let { rules += Rule(header.owner, null, it) }
        return rules
    }

    private fun parseNestedBody(lines: List<String>, header: Header): List<Rule>? {
        val rules = mutableListOf<Rule>()
        var section: String? = null
        var fallback = header.fallback
        for (raw in lines) {
            val code = codePart(raw)
            if (code.isEmpty() || code.startsWith("#")) continue
            when {
                code == "threads {" && section == null -> section = "threads"
                code == "processes {" && section == null -> section = "processes"
                code == "}" && section != null -> section = null
                section == "threads" -> splitAssignment(code)?.let { (name, cpus) ->
                    if (!validMemberName(name)) return null
                    rules += Rule(header.owner, name, cpus)
                } ?: return null
                section == "processes" -> splitAssignment(code)?.let { (name, cpus) ->
                    if (!validMemberName(name)) return null
                    rules += processRule(header.owner, name, cpus)
                } ?: return null
                else -> {
                    val assignment = splitAssignment(code) ?: return null
                    if (assignment.first != "fallback" || fallback != null) return null
                    fallback = assignment.second
                }
            }
        }
        if (section != null) return null
        fallback?.let { rules += Rule(header.owner, null, it) }
        return rules
    }

    private fun parseYamlHeader(raw: String): Header? {
        if (leadingSpaces(raw) != 0) return null
        val owner = codePart(raw).removeSuffix(":").trim()
        if (!codePart(raw).endsWith(':') || owner.isEmpty() || owner.any(Char::isWhitespace) ||
            owner == "threads" || owner == "processes"
        ) return null
        return Header(owner, null, Format.YAML)
    }

    private fun yamlEnd(lines: List<String>, start: Int): Int {
        var index = start + 1
        while (index < lines.size) {
            val code = codePart(lines[index])
            if (code.isNotEmpty() && !code.startsWith("#") && leadingSpaces(lines[index]) == 0) break
            index++
        }
        return index
    }

    private fun parseYamlBody(lines: List<String>, owner: String): List<Rule>? {
        val rules = mutableListOf<Rule>()
        var section: String? = null
        var fallback: String? = null
        for (raw in lines) {
            val code = codePart(raw)
            if (code.isEmpty() || code.startsWith("#")) continue
            when (code) {
                "threads:" -> {
                    if (leadingSpaces(raw) != 4) return null
                    section = "threads"
                    continue
                }
                "processes:" -> {
                    if (leadingSpaces(raw) != 4) return null
                    section = "processes"
                    continue
                }
            }
            val split = code.lastIndexOf(':')
            if (split <= 0) return null
            val name = code.substring(0, split).trim()
            val cpus = code.substring(split + 1).trim()
            if (name.isEmpty() || cpus.isEmpty()) return null
            when {
                section == "threads" && leadingSpaces(raw) == 8 && validMemberName(name) ->
                    rules += Rule(owner, name, cpus)
                section == "processes" && leadingSpaces(raw) == 8 && validMemberName(name) ->
                    rules += processRule(owner, name, cpus)
                name == "fallback" && leadingSpaces(raw) == 4 && fallback == null -> fallback = cpus
                else -> return null
            }
        }
        fallback?.let { rules += Rule(owner, null, it) }
        return rules
    }

    private fun parseStandardMember(owner: String, code: String): Rule? {
        val (name, cpus) = splitAssignment(code) ?: return null
        if (!validMemberName(name)) return null
        return if (name.startsWith(':') || name.startsWith("$owner:")) {
            processRule(owner, name, cpus)
        } else {
            Rule(owner, name, cpus)
        }
    }

    private fun parseTaggedMember(owner: String, code: String): Rule? {
        val (name, cpus) = splitAssignment(code) ?: return null
        return when {
            name == "fallback" -> Rule(owner, null, cpus)
            name.startsWith("thread:") && validMemberName(name.substring(7)) ->
                Rule(owner, name.substring(7), cpus)
            name.startsWith("process:") && validMemberName(name.substring(8)) ->
                processRule(owner, name.substring(8), cpus)
            else -> null
        }
    }

    private fun parseNaturalMember(owner: String, code: String): Rule? {
        return when {
            code.startsWith("thread ") -> splitAssignment(code.substring(7))?.let {
                if (!validMemberName(it.first)) return null
                Rule(owner, it.first, it.second)
            }
            code.startsWith("process ") -> splitAssignment(code.substring(8))?.let {
                if (!validMemberName(it.first)) return null
                processRule(owner, it.first, it.second)
            }
            else -> null
        }
    }

    private fun parseFunctionMember(owner: String, code: String): Rule? {
        val process = code.startsWith("process(")
        val prefix = if (process) "process(" else "thread("
        if (!code.startsWith(prefix) || !code.endsWith(')')) return null
        val values = splitFunctionArgs(code.substring(prefix.length, code.length - 1))
        if (values.size != 2 || values.any(String::isEmpty) || !validMemberName(values[0])) return null
        return if (process) processRule(owner, values[0], values[1]) else Rule(owner, values[0], values[1])
    }

    private fun processRule(owner: String, name: String, cpus: String): Rule {
        val child = when {
            name.startsWith("$owner:") -> name
            name.startsWith(':') -> owner + name
            else -> "$owner:$name"
        }
        return Rule(child, null, cpus)
    }

    private fun splitAssignment(code: String): Pair<String, String>? {
        val eq = code.indexOf('=')
        if (eq <= 0) return null
        val name = code.substring(0, eq).trim()
        val cpus = code.substring(eq + 1).trim()
        return if (name.isEmpty() || cpus.isEmpty()) null else name to cpus
    }

    private fun splitFunctionArgs(args: String): List<String> {
        val comma = args.lastIndexOf(',')
        return if (comma < 0) {
            listOf(args.trim())
        } else {
            listOf(args.substring(0, comma).trim(), args.substring(comma + 1).trim())
        }
    }

    private fun braceBlockEnd(lines: List<String>, start: Int): Int? {
        var depth = 0
        for (index in start until lines.size) {
            val code = codePart(lines[index])
            if (code.isEmpty() || code.startsWith("#")) continue
            depth += code.count { it == '{' }
            depth -= code.count { it == '}' }
            if (depth == 0) return index + 1
            if (depth < 0) return null
        }
        return null
    }

    private fun leadingSpaces(raw: String): Int = raw.length - raw.trimStart().length

    private fun validMemberName(name: String): Boolean {
        return name.isNotEmpty() && name.none { it == '{' || it == '}' || it == '=' }
    }

    private fun codePart(rawLine: String): String = rawLine.substringBefore("//").trim()

    private val NATURAL_HEADER = Regex("^app\\s+(\\S+)(?:\\s+fallback\\s+(\\S+))?\\s*\\{$")
}

File: native_daemon/daemon_rs/src/daemon_core/rule_syntax.rs
// applist.conf 的各种区块语法统一在这里展开成旧版规则，执行层只处理一种语义。
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct CanonicalRule {
    pub key: String,
    pub cpus: String,
}

pub struct CanonicalGroup {
    pub rules: Vec<CanonicalRule>,
    pub block: bool,
}

#[derive(Clone, Debug)]
pub struct BlockRange {
    pub owner: String,
    pub start_line: usize,
    pub end_line: usize,
}

struct ParsedDocument {
    groups: Vec<CanonicalGroup>,
    ranges: Vec<BlockRange>,
    valid: bool,
}

#[derive(Clone, Copy, Eq, PartialEq)]
enum BlockKind {
    Standard,
    Tagged,
    Nested,
    Natural,
    Function,
    Yaml,
}

struct Header {
    owner: String,
    fallback: Option<String>,
    kind: BlockKind,
    valid: bool,
}

pub fn parse_config_groups(text: &str) -> Vec<CanonicalGroup> {
    parse_document(text).groups
}

pub fn block_ranges(text: &str) -> Option<Vec<BlockRange>> {
    let document = parse_document(text);
    document.valid.then_some(document.ranges)
}

fn parse_document(text: &str) -> ParsedDocument {
    let lines = text.lines().collect::<Vec<_>>();
    let mut groups = Vec::new();
    let mut ranges = Vec::new();
    let mut valid = true;
    let mut index = 0;

    while index < lines.len() {
        let raw = lines[index];
        let code = code_part(raw);
        if code.is_empty() || code.starts_with('#') {
            index += 1;
            continue;
        }

        if let Some(header) = parse_yaml_header(raw) {
            let end = yaml_end(&lines, index);
            let parsed = parse_yaml_body(&lines[index + 1..end], &header.owner);
            valid &= parsed.is_some();
            groups.push(CanonicalGroup {
                rules: parsed.unwrap_or_default(),
                block: true,
            });
            ranges.push(BlockRange {
                owner: header.owner,
                start_line: index,
                end_line: end,
            });
            index = end;
            continue;
        }

        if let Some(mut header) = parse_brace_header(code) {
            let Some(end) = brace_block_end(&lines, index) else {
                valid = false;
                break;
            };
            let close = code_part(lines[end - 1]);
            let mut block_valid = header.valid;
            if header.kind == BlockKind::Standard {
                let tail = close.strip_prefix('}').map(str::trim).unwrap_or_default();
                if let Some(cpus) = tail.strip_prefix('=').map(str::trim) {
                    if cpus.is_empty() || header.fallback.is_some() {
                        block_valid = false;
                    } else {
                        header.fallback = Some(cpus.to_string());
                    }
                } else if !tail.is_empty() {
                    block_valid = false;
                }
            } else if close != "}" {
                block_valid = false;
            }
            let body = &lines[index + 1..end - 1];
            let parsed = parse_brace_body(body, &header);
            block_valid &= parsed.is_some();
            valid &= block_valid;
            groups.push(CanonicalGroup {
                rules: if block_valid {
                    parsed.unwrap_or_default()
                } else {
                    Vec::new()
                },
                block: true,
            });
            ranges.push(BlockRange {
                owner: header.owner,
                start_line: index,
                end_line: end,
            });
            index = end;
            continue;
        }

        let rules = parse_legacy_rule(code).into_iter().collect();
        groups.push(CanonicalGroup {
            rules,
            block: false,
        });
        index += 1;
    }

    ParsedDocument {
        groups,
        ranges,
        valid,
    }
}

fn parse_brace_header(code: &str) -> Option<Header> {
    if !code.ends_with('{') {
        return None;
    }
    let prefix = code[..code.len() - 1].trim();
    if prefix == "app" || prefix.starts_with("app ") {
        let rest = prefix.strip_prefix("app ").unwrap_or_default();
        let mut parts = rest.split_whitespace();
        let owner = parts.next().unwrap_or_default().to_string();
        let tail = parts.collect::<Vec<_>>();
        let fallback = match tail.as_slice() {
            [] => None,
            ["fallback", cpus] if !cpus.is_empty() => Some((*cpus).to_string()),
            _ => {
                return Some(Header {
                    owner,
                    fallback: None,
                    kind: BlockKind::Natural,
                    valid: false,
                })
            }
        };
        let valid = !owner.is_empty();
        return Some(Header {
            owner,
            fallback,
            kind: BlockKind::Natural,
            valid,
        });
    }
    if let Some(rest) = prefix.strip_prefix("app(") {
        let Some(args) = rest.strip_suffix(')') else {
            return Some(Header {
                owner: String::new(),
                fallback: None,
                kind: BlockKind::Function,
                valid: false,
            });
        };
        let values = args.split(',').map(str::trim).collect::<Vec<_>>();
        let owner = values.first()?.trim().to_string();
        let fallback = values.get(1).map(|v| v.trim().to_string());
        return Some(Header {
            owner,
            fallback,
            kind: BlockKind::Function,
            valid: values.len() <= 2 && values.iter().all(|value| !value.is_empty()),
        });
    }

    let (owner, fallback, kind) = if let Some((owner, cpus)) = prefix.split_once('=') {
        let owner = owner.trim().to_string();
        let cpus = cpus.trim();
        if cpus.is_empty() {
            (owner, None, BlockKind::Tagged)
        } else {
            (owner, Some(cpus.to_string()), BlockKind::Standard)
        }
    } else {
        (prefix.to_string(), None, BlockKind::Standard)
    };
    if owner.is_empty() || owner.contains(['{', '}']) {
        return None;
    }
    Some(Header {
        owner,
        fallback,
        kind,
        valid: true,
    })
}

fn parse_brace_body(lines: &[&str], header: &Header) -> Option<Vec<CanonicalRule>> {
    if header.kind == BlockKind::Nested
        || (header.kind == BlockKind::Tagged
            && lines
                .iter()
                .any(|line| matches!(code_part(line), "threads {" | "processes {")))
    {
        return parse_nested_body(lines, header);
    }

    let mut rules = Vec::new();
    let mut body_fallback = false;
    for raw in lines {
        let code = code_part(raw);
        if code.is_empty() || code.starts_with('#') {
            continue;
        }
        let rule = match header.kind {
            BlockKind::Standard => parse_standard_member(&header.owner, code),
            BlockKind::Tagged => parse_tagged_member(&header.owner, code),
            BlockKind::Natural => parse_natural_member(&header.owner, code),
            BlockKind::Function => parse_function_member(&header.owner, code),
            BlockKind::Nested | BlockKind::Yaml => None,
        }?;
        if header.kind == BlockKind::Tagged && rule.key == header.owner {
            if body_fallback || header.fallback.is_some() {
                return None;
            }
            body_fallback = true;
        }
        rules.push(rule);
    }
    if let Some(cpus) = header.fallback.as_ref() {
        rules.push(canonical(&header.owner, cpus));
    }
    Some(rules)
}

fn parse_nested_body(lines: &[&str], header: &Header) -> Option<Vec<CanonicalRule>> {
    let mut rules = Vec::new();
    let mut section: Option<&str> = None;
    let mut fallback = header.fallback.clone();
    for raw in lines {
        let code = code_part(raw);
        if code.is_empty() || code.starts_with('#') {
            continue;
        }
        match code {
            "threads {" => {
                if section.is_some() {
                    return None;
                }
                section = Some("threads");
            }
            "processes {" => {
                if section.is_some() {
                    return None;
                }
                section = Some("processes");
            }
            "}" => section = None,
            _ if section == Some("threads") => {
                let (name, cpus) = split_assignment(code)?;
                if !valid_member_name(name) {
                    return None;
                }
                rules.push(thread_rule(&header.owner, name, cpus));
            }
            _ if section == Some("processes") => {
                let (name, cpus) = split_assignment(code)?;
                if !valid_member_name(name) {
                    return None;
                }
                rules.push(process_rule(&header.owner, name, cpus));
            }
            _ => {
                let (name, cpus) = split_assignment(code)?;
                if name != "fallback" || fallback.is_some() {
                    return None;
                }
                fallback = Some(cpus.to_string());
            }
        }
    }
    if section.is_some() {
        return None;
    }
    if let Some(cpus) = fallback {
        rules.push(canonical(&header.owner, &cpus));
    }
    Some(rules)
}

fn parse_yaml_header(raw: &str) -> Option<Header> {
    if leading_spaces(raw) != 0 {
        return None;
    }
    let code = code_part(raw);
    let owner = code.strip_suffix(':')?.trim();
    if owner.is_empty()
        || owner.contains(char::is_whitespace)
        || owner == "threads"
        || owner == "processes"
    {
        return None;
    }
    Some(Header {
        owner: owner.to_string(),
        fallback: None,
        kind: BlockKind::Yaml,
        valid: true,
    })
}

fn yaml_end(lines: &[&str], start: usize) -> usize {
    let mut index = start + 1;
    while index < lines.len() {
        let raw = lines[index];
        let code = code_part(raw);
        if !code.is_empty() && !code.starts_with('#') && leading_spaces(raw) == 0 {
            break;
        }
        index += 1;
    }
    index
}

fn parse_yaml_body(lines: &[&str], owner: &str) -> Option<Vec<CanonicalRule>> {
    let mut rules = Vec::new();
    let mut section: Option<&str> = None;
    let mut fallback: Option<String> = None;
    for raw in lines {
        let code = code_part(raw);
        if code.is_empty() || code.starts_with('#') {
            continue;
        }
        if code == "threads:" {
            if leading_spaces(raw) != 4 {
                return None;
            }
            section = Some("threads");
            continue;
        }
        if code == "processes:" {
            if leading_spaces(raw) != 4 {
                return None;
            }
            section = Some("processes");
            continue;
        }
        let (name, cpus) = code.rsplit_once(':')?;
        let name = name.trim();
        let cpus = cpus.trim();
        if name.is_empty() || cpus.is_empty() {
            return None;
        }
        match section {
            Some("threads") if leading_spaces(raw) == 8 && valid_member_name(name) => {
                rules.push(thread_rule(owner, name, cpus))
            }
            Some("processes") if leading_spaces(raw) == 8 && valid_member_name(name) => {
                rules.push(process_rule(owner, name, cpus))
            }
            _ if name == "fallback" && leading_spaces(raw) == 4 && fallback.is_none() => {
                fallback = Some(cpus.to_string())
            }
            _ => return None,
        }
    }
    if let Some(cpus) = fallback {
        rules.push(canonical(owner, &cpus));
    }
    Some(rules)
}

fn parse_standard_member(owner: &str, code: &str) -> Option<CanonicalRule> {
    let (name, cpus) = split_assignment(code)?;
    if !valid_member_name(name) {
        return None;
    }
    if name.starts_with(':') || name.starts_with(&format!("{owner}:")) {
        Some(process_rule(owner, name, cpus))
    } else {
        Some(thread_rule(owner, name, cpus))
    }
}

fn parse_tagged_member(owner: &str, code: &str) -> Option<CanonicalRule> {
    let (name, cpus) = split_assignment(code)?;
    if name == "fallback" {
        return Some(canonical(owner, cpus));
    }
    if let Some(thread) = name.strip_prefix("thread:") {
        return valid_member_name(thread).then(|| thread_rule(owner, thread, cpus));
    }
    let process = name.strip_prefix("process:")?;
    valid_member_name(process).then(|| process_rule(owner, process, cpus))
}

fn parse_natural_member(owner: &str, code: &str) -> Option<CanonicalRule> {
    if let Some(rest) = code.strip_prefix("thread ") {
        let (name, cpus) = split_assignment(rest)?;
        if !valid_member_name(name) {
            return None;
        }
        return Some(thread_rule(owner, name, cpus));
    }
    let rest = code.strip_prefix("process ")?;
    let (name, cpus) = split_assignment(rest)?;
    if !valid_member_name(name) {
        return None;
    }
    Some(process_rule(owner, name, cpus))
}

fn parse_function_member(owner: &str, code: &str) -> Option<CanonicalRule> {
    let (process, args) = if let Some(args) = code
        .strip_prefix("thread(")
        .and_then(|v| v.strip_suffix(')'))
    {
        (false, args)
    } else {
        (true, code.strip_prefix("process(")?.strip_suffix(')')?)
    };
    let values = split_function_args(args);
    if values.len() != 2 {
        return None;
    }
    let name = values[0].trim();
    let cpus = values[1].trim();
    if !valid_member_name(name) || cpus.is_empty() {
        return None;
    }
    Some(if process {
        process_rule(owner, name, cpus)
    } else {
        thread_rule(owner, name, cpus)
    })
}

fn parse_legacy_rule(code: &str) -> Option<CanonicalRule> {
    let (key, cpus) = split_assignment(code)?;
    Some(canonical(key, cpus))
}

fn split_assignment(code: &str) -> Option<(&str, &str)> {
    let (name, cpus) = code.split_once('=')?;
    let name = name.trim();
    let cpus = cpus.trim();
    (!name.is_empty() && !cpus.is_empty()).then_some((name, cpus))
}

fn split_function_args(args: &str) -> Vec<&str> {
    match args.rsplit_once(',') {
        Some((left, right)) => vec![left.trim(), right.trim()],
        None => vec![args.trim()],
    }
}

fn thread_rule(owner: &str, name: &str, cpus: &str) -> CanonicalRule {
    canonical(&format!("{owner}{{{name}}}"), cpus)
}

fn process_rule(owner: &str, name: &str, cpus: &str) -> CanonicalRule {
    let child = if name.starts_with(&format!("{owner}:")) {
        name.to_string()
    } else if name.starts_with(':') {
        format!("{owner}{name}")
    } else {
        format!("{owner}:{name}")
    };
    canonical(&child, cpus)
}

fn canonical(key: &str, cpus: &str) -> CanonicalRule {
    CanonicalRule {
        key: key.to_string(),
        cpus: cpus.to_string(),
    }
}

fn brace_block_end(lines: &[&str], start: usize) -> Option<usize> {
    let mut depth = 0isize;
    for (index, raw) in lines.iter().enumerate().skip(start) {
        let code = code_part(raw);
        if code.is_empty() || code.starts_with('#') {
            continue;
        }
        depth += code.chars().filter(|ch| *ch == '{').count() as isize;
        depth -= code.chars().filter(|ch| *ch == '}').count() as isize;
        if depth == 0 {
            return Some(index + 1);
        }
        if depth < 0 {
            return None;
        }
    }
    None
}

fn leading_spaces(raw: &str) -> usize {
    raw.len() - raw.trim_start().len()
}

fn valid_member_name(name: &str) -> bool {
    !name.is_empty() && !name.contains(['{', '}', '='])
}

fn code_part(raw: &str) -> &str {
    raw.split_once("//").map_or(raw, |(code, _)| code).trim()
}
